flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -66,6 +76,43 @@ public void init() {
 			operatorChain);
 	}
 
+	@Override
+	protected void beforeInvoke() throws Exception {
```

Comment:
We are creating `input` and the `inputProcessor` twice now, once in `init()` and once on `beforeInvoke()`. Can we add this code here to init to avoid creating this twice? I think creating it twice may have some unexpected side effects in certain cases.

Reply:
Good catch. I forgot to delete the `Input` / `InputProcessor` creation in `init()`.
We cannot add the code here to `init()` because the `SourceReader` was created after the `init()` invocation.

Reply:
I think just deleting init() can be dangerous. It means we create the input processors much later in the setup sequence (after the `beforeInvoke()` rather than early during that), which changes the setup can lead to surprises down the way.
It may be that some part of the input sequence rely on the processors being present.

The branch I referenced below has a way to create the reader earlier and thus maintain the SteamTask's setup sequence.